frontend/apps/reader/modules/readerlink.lua

```diff
@@ -687,137 +820,30 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
-            if util.pathExists(epub_path) then
-                epub_fullpath = epub_path
-            end
-        end
-        -- or in wikipedia save directory
-        if not epub_fullpath then
-            local dir = G_reader_settings:readSetting("wikipedia_save_dir")
-            if not dir then dir = G_reader_settings:readSetting("home_dir") end
-            if not dir then dir = require("apps/filemanager/filemanagerutil").getDefaultDir() end
-            if dir then
-                local safe_filename = util.getSafeFilename(epub_filename, dir):gsub("_", " ")
-                local epub_path = dir .. "/" .. safe_filename
-                if util.pathExists(epub_path) then
-                    epub_fullpath = epub_path
+    local default_title =  T(_("External link:\n\n%1"), BD.url(link_url))
+    local title = default_title
+
+    for _, fn_button in pairs(self._external_buttons) do
+        local button = fn_button(self, link_url)
+        if button.show_in_dialog_func then
+            local show, button_title = button.show_in_dialog_func()
+            if show and button_title then
+                if title == default_title then
+                    -- The default title is replaced by the first non-default button title.
+                    title = button_title
+                else
+                    -- Every other button title value is appended to the title.
+                    title = title .. "\n\n" .. button_title
                 end
             end
         end
-        text = T(_("Would you like to read this Wikipedia %1 article?\n\n%2\n"), wiki_lang:upper(), wiki_page:gsub("_", " "))
-        add_button({
-            text = _("Read online"),
-            callback = function()
-                UIManager:nextTick(function()
-                    UIManager:close(dialog)
-                    self.ui:handleEvent(Event:new("LookupWikipedia", wiki_page, true, false, true, wiki_lang))
-                end)
-            end,
-        })
-        if epub_fullpath then
-            text = T("%1\n%2", text, _("This article has previously been saved as EPUB. You may wish to read the saved EPUB instead."))
-            add_button({
-                text = _("Read EPUB"),
-                callback = function()
-                    UIManager:scheduleIn(0.1, function()
-                        UIManager:close(dialog)
-                        self.ui:switchDocument(epub_fullpath)
-                    end)
-                end,
-            })
-        end
-    else
-        if #alt_handlers_buttons == 0 then
-            -- No external link handler
-            return false
-        end
-        text = T(_("External link:\n\n%1"), BD.url(link_url))
     end
 
-    -- Add all alternative handlers buttons
-    for __, button in ipairs(alt_handlers_buttons) do
-        add_button(button)
-    end
-    dialog = ButtonDialogTitle:new{
-        title = text,
-        buttons = buttons,
+    self.dialog = ButtonDialogTitle:new{
```

Comment:
Name this `self.external_link_dialog`, it makes it easier to grasp when reading all the little `function(this, link_uri)` where it's already not intuitive to get what `this` is.

Reply:
`self.dialog` is used elsewhere in this module, and I'm not sure I want to disturb or make sense of those usages... In those contexts, `self.dialog` actually isn't always an external link (In once place it's a Footnote widget).

Reply:
Oh, but that's not the same!
`self.dialog` is the parent of readerlink, so readerui, and it is used when refreshing (in many widgets, we pass the real parent widget (the widget UIManager:show()n) down to each little widget that makes this real widget via self.dialog. So, when a button needs to refresh itself, it requests self.dialog to repaint/refresh this little region (so, we know we don't need to repaint/refresh all the other widgets, ie the book, the footer).

So, one reason to really give it another name (previously, it was a `local dialog`, nothing to do with `self.dialog`.